lib/TaskMonitoring/include/TaskMonitoring/shared_reference.h

```diff
@@ -0,0 +1,115 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2024 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Business Source License 1.1 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     https://github.com/arangodb/arangodb/blob/devel/LICENSE
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Volmer
+////////////////////////////////////////////////////////////////////////////////
+#pragma once
+
+#include <cstdlib>
+#include <functional>
+#include <atomic>
+
+/**
+   Reference counting wrapper for a resource
+
+   Destroys itself and calls a custom cleanup function on the resource when the
+   reference count decrements to zero.
+   This struct provides a similar interface and behaves similarly to a
+   shared_ptr, but it provides a ref_count function that uses a release memory
+   order.
+ */
+template<typename T>
+struct Shared {
```

Comment:
This looks correct, but could you educate me why this could not be solved with std::shared_ptr ?
It looks pretty close to it.

Reply:
Now that you mention it, I thought I need it because I need the number of references to the shared reference in `mark_finished_nodes_for_deletion`, which is called when a `NodeReference` is destructed (defined in in the `Task` constructor). `std::shared_ptr` has the function `use_count`, but cppreference says it possibly uses relaxed memory ordering. I thought that I need acquire memory order (I actually used for some reason the wrong - release - one here which does not make any sense, it should have been acquire instead) here in order to synchronize correctly with decrementing and incrementing the shared reference counter, but now that I think about it, it is totally fine to use relaxed ordering. I added some comments why this is ok and use shared ptr now. Thanks :)